onnx/runtime/op_run.py

```diff
@@ -0,0 +1,463 @@
+            ) from e
+        self._log("-- done %s.run -> %d outputs", self.__class__.__name__, len(res))
+        if not isinstance(res, tuple):
+            raise TypeError(
+                f"Method '_run' of class {self.__class__.__name__!r} does not return a tuple but {type(res)}."
+            )
+        if len(res) == 0:
+            raise ValueError(
+                f"Method '_run' of class {self.__class__.__name__!r} does not return any result."
+            )
+        if any(map(lambda t: isinstance(t, tuple), res)):
+            dtypes = [type(t) for t in res]
+            raise TypeError(
+                f"Method '_run' of class {self.__class__.__name__!r} must not return a tuple: {dtypes}."
+            )
+        return res
+
+    @classmethod
+    def make_node(
+        cls,
+        n_inputs: Optional[int] = None,
+        n_outputs: Optional[int] = None,
+        **kwargs: Any,
+    ) -> NodeProto:  # type: ignore
+        """
+        Creates an ONNX node for this class based on the given information.
+
+        :param n_inputs: number of inputs (default is defined by the operator schema)
+        :param n_outputs: number of outputs (default is defined by the operator schema)
+        :param verbose: verbosity
+        :param kwargs: node attributes
+        :return: NodeProto
+        """
+        domain = cls.op_domain
+        schema = None
+        if n_inputs is None:
+            if schema is None:
+                schema = get_schema(cls.__name__, onnx_opset_version(), domain)
+            n_inputs = schema.min_input
+        if n_outputs is None:
+            if schema is None:
+                schema = get_schema(cls.__name__, onnx_opset_version(), domain)
+            n_outputs = schema.min_output
+
+        names_in = [f"x{i}" for i in range(n_inputs)]
+        names_out = [f"y{i}" for i in range(n_outputs)]
+        node = make_node(cls.__name__, names_in, names_out, **kwargs)
+        return node
+
+    @classmethod
+    def create(
+        cls,
+        n_inputs: Optional[int] = None,
+        n_outputs: Optional[int] = None,
+        verbose: int = 0,
+        **kwargs: Any,
+    ) -> Any:
+        """
+        Instantiates this class based on the given information.
+
+        :param n_inputs: number of inputs (default is defined by the operator schema)
+        :param n_outputs: number of outputs (default is defined by the operator schema)
+        :param verbose: verbosity
+        :param kwargs: node attributes
+        :return: NodeProto
+        """
+
+        def log_function(pattern: str, *args: Any) -> None:
+            if verbose > 1:
+                print(pattern % tuple(args))
+
+        node = cls.make_node(n_inputs, n_outputs, **kwargs)
+        run_params = dict(verbose=verbose, log=log_function)
+        cl = cls(node, run_params)
+        return cl
+
+    @classmethod
+    def eval(
+        cls,
+        *args: List[Any],
+        n_outputs: Optional[int] = None,
```

Comment:
nit: may be useful to give it a name like `__n_outputs` to avoid conflicting with any ONNX op attribute-name. Same for `verbose`.

Reply:
Is it possible to pass in a dict for attributes instead?

Reply:
If possible, I'd like to keep verbose as it is. I doubt an operator would use it as this parameter means the same thing accross all packages. Same goes for n_outputs. The meaning of the parameter is quite clear. If an operator uses it in its definition, it should have the same meaning. I don't mind using an attribute instead of `**kwargs`. Any particular reason? `make_node` is also using `**kwargs`.

Reply:
> I don't mind using an attribute instead of `**kwargs`. Any particular reason? `make_node` is also using `**kwargs`.

Just to avoid naming conflicts if there is a need.